engines/private/funcs.cpp

```diff
@@ -0,0 +1,784 @@
+            if (strcmp(c.u.str, "\"REMOVE\"") == 0) {
+                v1.u.sym->u.val = 0;
+                g_private->inventory.remove(*bmp);
+            } else {
+                v1.u.sym->u.val = 1;
+                g_private->inventory.push_back(*bmp);
+            }
+        } else {
+            g_private->inventory.push_back(*bmp);
+        }
+        if (v2.type == NAME)
+            v2.u.sym->u.val = 1;
+    }
+}
+
+void fSetFlag(ArgArray args) {
+    assert(args.size() == 2);
+    assert(args[0].type == NAME && args[1].type == NUM);
+    debugC(1, kPrivateDebugScript, "SetFlag(%s, %d)", args[0].u.sym->name->c_str(), args[1].u.val);
+    args[0].u.sym->u.val = args[1].u.val;
+}
+
+void fExit(ArgArray args) {
+    // assert types
+    assert(args[2].type == RECT || args[2].type == NAME);
+    debugC(1, kPrivateDebugScript, "Exit(%d %d %d)", args[0].type, args[1].type, args[2].type); //, args[0].u.str, args[1].u.sym->name->c_str(), "RECT");
+    ExitInfo *e = (ExitInfo *)malloc(sizeof(ExitInfo));
+
+    if (args[0].type == NUM && args[0].u.val == 0)
+        e->nextSetting = NULL;
+    else
+        e->nextSetting = new Common::String(args[0].u.str);
+
+    if (args[1].type == NUM && args[1].u.val == 0)
+        e->cursor = NULL;
+    else
+        e->cursor = args[1].u.sym->name;
+
+    if (args[2].type == NAME) {
+        assert(args[2].u.sym->type == RECT);
+        args[2].u.rect = args[2].u.sym->u.rect;
+    }
+
+    e->rect = args[2].u.rect;
+    g_private->_exits.push_front(*e);
+}
+
+void fSetModifiedFlag(ArgArray args) {
+    // assert types
+    debugC(1, kPrivateDebugScript, "SetModifiedFlag(%d)", args[0].u.val);
+    g_private->_modified = args[0].u.val != 0;
+}
+
+void fPaperShuffleSound(ArgArray args) {
+    assert(args.size() == 0);
+    debugC(1, kPrivateDebugScript, "PaperShuffleSound()");
+    Common::String *s = g_private->getPaperShuffleSound();
+    g_private->playSound(*s, 1, false, false);
+}
+
+void fSoundEffect(ArgArray args) {
+    // assert types
+    debugC(1, kPrivateDebugScript, "SoundEffect(%s)", args[0].u.str);
+    if (strcmp("\"\"", args[0].u.str) != 0) {
+        Common::String *s = new Common::String(args[0].u.str);
+        g_private->playSound(*s, 1, false, false);
+        //assert(0);
+    } else {
+        g_private->stopSound(true);
+    }
+}
+
+void fSound(ArgArray args) {
+    // assert types
+    debugC(1, kPrivateDebugScript, "Sound(%s)", args[0].u.str);
+    if (args.size() == 4) {
+        bool b1 = args[1].u.val != 0;
+        bool b2 = args[2].u.val != 0;
+        int c = args[3].u.val;
+
+        if (!b1 && !b2 && c == 1) {
+            g_private->stopSound(true);
+        } else if (!b1 && !b2 && c == 2) {
+            g_private->stopSound(false);
+        } else
+            assert(0);
+    }
+
+    if (strcmp("\"\"", args[0].u.str) != 0) {
+        Common::String *s = new Common::String(args[0].u.str);
+        g_private->playSound(*s, 1, false, false);
```

Comment:
I think you may change the `playSound` prototype to accept `Common::String str&`, e.g. pass it by reference, because you're not changing it in there. This would simplify things and remove this memory leak.

Reply:
Fixed.